Add "Optimizing LCP using Signed Exchanges" post. (#2613)
GoogleChrome/developer.chrome.com: site/en/blog/optimizing-lcp-using-signed-exchanges/index.md

---
layout: 'layouts/blog-post.njk'
title: Optimizing LCP using Signed Exchanges
description: >
  Signed exchanges are a means for web developers to drastically improve page loading speed coming from SXG referrers such as Google Search. They can provide benefit simply from being enabled by a checkbox, but you can take additional steps to get the most out of them.
subhead: >
  How to measure and optimize signed exchanges to get the most improvement out of them.

> date: 2022-04-15

Reviewer: ```suggestion
date: 2022-04-18
```

> 
> authors:
>   - twifkak
> 
> tags:
>   - performance

Reviewer: ```suggestion
authors:
  - twifkak
tags:
  - performance
```

> 
> # Optional
> # Hero images should be at least 1920 x 960
> # https://developer.chrome.com/docs/handbook/how-to/add-media/
> # hero: 'image/BrQidfK9jaQyIHwdw91aVpkPiib2/EnMzOm0mBytBA3AzlCG6.png'
> 
> # Optional
> # You can provide an optional cropping of your hero image to be used as a
> # thumbnail. Note the alt text will be the same for both the thumbnail and
> # the hero.
> # Thumbnail images should be at least 828 x 416
> # thumbnail: 'image/BrQidfK9jaQyIHwdw91aVpkPiib2/EnMzOm0mBytBA3AzlCG6.png'
> 
> # Required if there is a hero or social image
> # alt: An alternative text description of your hero or social image.

Reviewer: ```suggestion
```

Reviewer: Just removing some boilerplate here. If you need to add hero images, let me know. I can edit them. Otherwise, we're fine to publish without them.

---

> 
> <!--
> - Only use h2-h4s in markdown.
> - Make sure all of your images have `alt` text.
> - Code blocks should use triple backticks and include a language name.
> -->

Reviewer: ```suggestion
```

Signed exchanges (SXGs) are a means for web developers to improve their page speed—mainly [Largest Contentful Paint (LCP)](https://web.dev/lcp/)—by enabling other sites (currently Google Search) to [prefetch their pages](https://developers.google.com/search/docs/advanced/experience/signed-exchange) before the user clicks on the link.

{% Aside %}
SXGs are potentially useful for other use cases not covered in this article. For instance, they could enable connection resilience: a chat app could pre-download an SXG of a link sent to the user, so they could later view it while offline. See this larger list of [potential use cases](https://wicg.github.io/webpackage/draft-yasskin-wpack-use-cases.html).
{% endAside %}

It is possible to make web pages that completely eliminate the network from the critical path to rendering the page, when prefetched! On 4G, this page load [goes from 6s to 0.9s](https://www.webpagetest.org/video/compare.php?tests=220329_BiDc1V_FJG,220329_AiDcB7_HB0&medianMetric=LCP). What remains is bound mostly by CPU:

> {% Video src="video/rULxC7pPw3PFS4o9xr7v8isFmCv1/5IRBMSsCZwihnVIVK4gS.mp4",
>          controls=true,
>          poster="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/4ttTTum2aZuNZL04vYmI.jpg" %}

Reviewer: ```suggestion
<figure>
  {% Video src="video/rULxC7pPw3PFS4o9xr7v8isFmCv1/5IRBMSsCZwihnVIVK4gS.mp4", controls=true, poster="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/4ttTTum2aZuNZL04vYmI.jpg" %}
</figure>
```

Reviewer: Since these are figures, they should be encased in a `<figure>` element for those on assistive technology.

In the past couple of months since Cloudflare's launch, I've been reading and responding to questions on [various](https://support.google.com/webmasters/threads?hl=en&thread_filter=(%27%22signed%20exchange%22%27)) [forums](https://community.cloudflare.com/t/automatic-signed-exchanges-sxgs-beta-launch/312634) and advising sites on how to make sure they're getting the most out of their signed exchange deployments. In the process, I think I've learned enough to begin offering some common advice for frequently asked questions.

Let's dig in.

## Analyze

One place to start is to use a lab tool like WebPageTest like shown above, so you can compare waterfalls—and LCP—with and without SXG prefetch.

{% Aside 'caution' %}
[Lab results may differ](https://web.dev/lab-and-field-data-differences/) from real user experiences, but they are a useful tool to diagnose what needs optimization and iterate quickly. We'll cover how to measure the resultant field data in the [Measure](#measure) section.
{% endAside %}

Generate a test without SXG in the standard way: go to [WebPageTest](https://www.webpagetest.org/) and enter a URL.

>   - Consider doing this while signed in, so you can more easily generate a comparison.
>   - Go to "Advanced Configuration" and choose a set of options there, rather than "Simple Configuration". (This makes it easy to assure that you have the same set of parameters as the "with SXG" test, which will use Advanced.)
>   - Under Browser, choose an Android phone under "Chrome Device Emulation". (This is also for parity with the "with SXG" test. As of writing, Google Search only supports SXG on Android.)
>   - Everything else is up to you. It may be helpful to increase "Number of Tests to Run" to 7, and pick a 4G connection.

Reviewer: ```suggestion
- Consider doing this while signed in, so you can more easily generate a comparison.
- Go to "Advanced Configuration" and choose a set of options there, rather than "Simple Configuration". (This makes it easy to assure that you have the same set of parameters as the "with SXG" test, which will use Advanced.)
- Under Browser, choose an Android phone under "Chrome Device Emulation". (This is also for parity with the "with SXG" test. As of writing, Google Search only supports SXG on Android.)
- Everything else is up to you. It may be helpful to increase "Number of Tests to Run" to 7, and pick a 4G connection.
```

Generate a test with SXG by starting with the same steps as above, but then go to the Script tab and enter a [WebPageTest script](https://docs.webpagetest.org/scripting/) like the following. This script overrides the URL entered in the main input field:

```clike
// Disable log collection for the first step. We only want the waterfall for the target navigation.
logData 0

// Visit a search result page that includes your page.
navigate https://google.com/search?q=site%3Asigned-exchange-testing.dev+image

// Wait for the prefetch to succeed.
sleep 10

// Re-enable log collection.
logData 1

// Navigate to the prefetched SXG on the Google SXG Cache.
navigate https://signed--exchange--testing-dev.webpkgcache.com/doc/-/s/signed-exchange-testing.dev/sxgs/valid-image-subresource.html
```

> If your page doesn't appear in any Google Search results yet, you can use [this prefetch page](https://signed-exchange-testing.dev/prefetch/) to generate a pretend search results page for this purpose.
> 
> 

Reviewer: ```suggestion
If your page doesn't appear in any Google Search results yet, you can use [this prefetch page](https://signed-exchange-testing.dev/prefetch/) to generate a pretend search results page for this purpose.

```

{% Aside %}

> Note that when Google Search instructs a user's browser to prefetch your SXG, it does so via a URL on webpkgcache.com in order to [avoid revealing the user's intent](https://wicg.github.io/webpackage/draft-yasskin-wpack-use-cases.html#name-privacy-preserving-prefetch) before they navigate. These URLs don't affect the display or behavior of the page, because the browser trusts the original URL that has been signed into the SXG.

Reviewer: ```suggestion
When Google Search instructs a user's browser to prefetch your SXG, it does so via a URL on webpkgcache.com in order to [avoid revealing the user's intent](https://wicg.github.io/webpackage/draft-yasskin-wpack-use-cases.html#name-privacy-preserving-prefetch) before they navigate. These URLs don't affect the display or behavior of the page, because the browser trusts the original URL that has been signed into the SXG.
```

{% endAside %}

The easiest way to determine the webpkgcache.com URL is to visit your page using the [SXG Validator Chrome extension](https://chrome.google.com/webstore/detail/sxg-validator/hiijcdgcphjeljafieaejfhodfbpmgoe), and click the extension icon to see the cache URL:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/DbkcylGpILjOzOnxREXl.png",
>        alt="SXG Validator showing cache information including URL",
>        width="506", height="229" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/DbkcylGpILjOzOnxREXl.png", alt="SXG Validator showing cache information including URL", width="506", height="229" %}
</figure>
```

Once these tests are complete, go to [Test History](https://app.webpagetest.org/ui/wpt/testhistory), select the two tests, and click Compare:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/ElSV7nAWCWV0Yjf7VC2o.png",
>        alt="Test History with two tests checked and the Compare button highlighted",
>        width="509", height="202" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/ElSV7nAWCWV0Yjf7VC2o.png", alt="Test History with two tests checked and the Compare button highlighted", width="509", height="202" %}
</figure>
```

Append `?medianMetric=LCP` to the compare URL so that the run with median LCP is selected for each side of the comparison. (The default is median by Speed Index.)

To compare waterfalls, expand the Waterfall Opacity section and drag the slider. To view the video, click Adjust Filmstrip Settings, scroll down inside that dialog, and click View Video.

## Debug

If the WebPageTest is showing that the HTML is being prefetched, then it has succeeded in all the steps of the pipeline; you may skip to the [Optimize](#optimize) section to learn how to further improve LCP. Otherwise, you'll need to find out where in the pipeline it failed and why; see the following sections for how.

### Publishing

Make sure your pages are being generated as signed exchanges.

{% Aside %}

> Your site will not serve SXG to users, even on SXG-capable browsers. They're meant for SXG-capable crawlers, to enable prefetch. (Serving SXGs to users who visit the site directly only adds overhead over a perfectly cromulent web page.) Thus, to debug, you need to pretend to be a crawler.

Reviewer: I missed this during edit: I get the reference with the word "cromulent"—and I love that episode, btw 😂—but our content is consumed by an international audience, and with "cromulent" being a made-up word from a primarily English-language television show, I wonder if it won't just cause confusion. I'd consider changing this to "perfectly acceptable" or some such equivalent.

Author: Thanks! Good suggestion; fixed.

{% endAside %}

The easiest way is to use the SXG Validator Chrome extension:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/zUqSkRSLNkF26gjyymEo.png",
>        alt="SXG Validator showing a check mark (✅) and a Content Type of application/signed-exchange;v=b3",
>        width="600", height="281" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/zUqSkRSLNkF26gjyymEo.png", alt="SXG Validator showing a check mark (✅) and a Content Type of application/signed-exchange;v=b3", width="600", height="281" %}
</figure>
```

> The extension fetches the current URL again, but with an Accept header that says it prefers the SXG version. If you see a check mark (✅), that means an SXG was returned; you can skip to the [Indexing](#indexing) section.

Reviewer: ```suggestion
The extension fetches the current URL again, but with an `Accept` request header that says it prefers the SXG version. If you see a check mark (✅), that means an SXG was returned; you can skip to the [Indexing](#indexing) section.
```

If you see a cross mark (❌), that means an SXG wasn't returned:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/4T3la9HtoRDz8SK7lWGi.png",
>        alt="SXG Validator showing a cross mark (❌) and a Content Type of text/html",
>        width="483", height="162" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/4T3la9HtoRDz8SK7lWGi.png", alt="SXG Validator showing a cross mark (❌) and a Content Type of text/html", width="483", height="162" %}
</figure>
```

> If Cloudflare ASX is enabled, then the most likely reason for a cross mark (❌) is because a cache control response header prevents it. ASX looks at headers with the following names:
>   - `Cache-Control`
>   - `CDN-Cache-Control`
>   - `Surrogate-Control`
>   - `Cloudflare-CDN-Cache-Control`
> 
> If any of these headers contains any of the following header values, it will prevent an SXG from being generated:
>   - `private`
>   - `no-store`
>   - `no-cache`
>   - `s-maxage` less than 120, if present, or else `max-age` less than 120

Reviewer: ```suggestion
If Cloudflare ASX is enabled, then the most likely reason for a cross mark (❌) is because a cache control response header prevents it. ASX looks at headers with the following names:
- `Cache-Control`
- `CDN-Cache-Control`
- `Surrogate-Control`
- `Cloudflare-CDN-Cache-Control`

If any of these headers contains any of the following header values, it will prevent an SXG from being generated:
- `private`
- `no-store`
- `no-cache`
- `s-maxage` less than 120, if present, or else `max-age` less than 120
```

{% Aside %}
In the future, you will be able to designate an SXG as only for new visitors, using `Vary: Cookie`. This is [implemented in Chromium](https://crbug.com/1250532), and will be enabled in Google Search once that has rolled out to enough people.
{% endAside %}

Alternatively to the Chrome extension, you can use a tool like `curl`:

```bash
curl -siH "Accept: application/signed-exchange;v=b3" $URL | less
```

or [`dump-signedexchange`](https://github.com/WICG/webpackage/blob/main/go/signedexchange/README.md#dump-a-signed-exchange-file):

```bash
dump-signedexchange -verify -uri $URL
```

### Indexing

Make sure your SXGs are successfully [indexed](https://support.google.com/webmasters/answer/7645831) by Google Search. Test if Google Search is linking to webpkgcache.com's copy of your page, by looking in DevTools. Since Google Search currently only supports SXG for Android OSes, you may need to enable mobile-emulation mode and select an Android device:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/Qu2qWl2LES3N8tiqnskP.png",
>        alt="Google Search results with DevTools showing an <a> that points to webpkgcache.com",
>        width="800", height="635" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/Qu2qWl2LES3N8tiqnskP.png", alt="Google Search results with DevTools showing an <a> that points to webpkgcache.com", width="800", height="635" %}
</figure>
```

If Google Search thinks the user is likely to click on the result, it will also prefetch it:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/uhchyec7uo4rgcQsteaE.png",
>        alt="Google Search results with DevTools showing a <link rel=prefetch> for webpkgcache.com",
>        width="800", height="147" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/uhchyec7uo4rgcQsteaE.png", alt="Google Search results with DevTools showing a <link rel=prefetch> for webpkgcache.com", width="800", height="147" %}
</figure>
```

The `<link>` tag instructs the browser to download the SXG into its prefetch cache. When the user clicks on the `<a>` tag, the browser will use that cached SXG to render the page.

You can also see evidence of this by going to the Network tab in DevTools and searching for URLs containing `webpkgcache`.

If this is happening, this means that Google Search indexing of the signed exchange is working. You can skip forward to the [Ingestion](#ingestion) section.

> Otherwise, it could be that Google hasn't recrawled the page yet since you enabled SXG. Try the Search Console URL Inspection Tool:

Reviewer: ```suggestion
Otherwise, it could be that Google hasn't recrawled the page yet since you enabled SXG. Try the [Google Search Console](https://search.google.com/search-console/about) URL Inspection Tool:
```

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/vaPuQ3k315xVxdn9elPe.png",
>        alt="Search Console URL Inspection tool, clicking View Crawled Page and then More Info",
>        width="800", height="302" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/vaPuQ3k315xVxdn9elPe.png", alt="Search Console URL Inspection tool, clicking View Crawled Page and then More Info", width="800", height="302" %}
</figure>
```

The presence of a `digest: mi-sha256-03=...` header indicates that Google successfully crawled the SXG version.

If a `digest` header is not present, this could be an indication that an SXG was not served to Googlebot, or that the index hasn't been updated yet.

If an SXG is successfully crawled, but it still isn't being linked to, then it may be a failure to meet SXG cache requirements. These are covered in the next section.

### Ingestion

When Google Search indexes an SXG, it sends its copy to the Google SXG Cache, which validates it against the [cache requirements](https://github.com/google/webpackager/blob/main/docs/cache_requirements.md). The Chrome extension shows the result:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/pB76K9ocvcooeZIpVxIq.png",
>        alt="SXG Validator showing a check mark (✅) and no warning message",
>        width="390", height="159" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/pB76K9ocvcooeZIpVxIq.png", alt="SXG Validator showing a check mark (✅) and no warning message", width="390", height="159" %}
</figure>
```

If you see a check mark (✅), then you can skip ahead to [Optimize](#optimize).

If it fails to meet the requirements, a warning message indicates why:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/fdNg8KIIXGsfSBpVCaVB.png",
>        alt='SXG Validator showing a cross mark (❌) and a warning message saying "error… parsing signature"',
>        width="359", height="258" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/fdNg8KIIXGsfSBpVCaVB.png", alt='SXG Validator showing a cross mark (❌) and a warning message saying "error… parsing signature"', width="359", height="258" %}
</figure>
```

In this case, the page will work just as it did before enabling SXG. Google will link to the page on its original host without an SXG prefetch.

In the event that the cached copy has expired and is being re-fetched in the background, you will see an hourglass (⌛):

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/BG0EYkGJOH9U391725he.png",
>        alt="SXG Validator showing an hourglass (⌛) and no warning message",
>        width="377", height="188" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/BG0EYkGJOH9U391725he.png", alt="SXG Validator showing an hourglass (⌛) and no warning message", width="377", height="188" %}
</figure>
```

The Google developer document on SXG also has instructions for [querying the cache manually](https://developers.google.com/search/docs/advanced/experience/signed-exchange#debug-the-google-sxg-cache).

## Optimize

If the extension shows all check marks (✅), you have a SXG that can be served to users! But there are some things you can do to optimize the web page so that you get the most LCP improvement from SXG.

### max-age

When SXGs expire, the Google SXG Cache will fetch a new copy in the background. While waiting for that fetch, users are directed to the page on its original host, which is not prefetched. The longer your `Cache-Control: max-age`, the less often this background fetch happens, and thus the more often that LCP can be reduced by prefetch.

> This is a tradeoff between performance and freshness, and the cache allows site owners to provide SXGs with a max-age anywhere between 2 minutes and 7 days, to fit each page's particular needs. Anecdotally, we find that:
>   - `max-age=86400` (1 day) or longer works well for performance
>   - `max-age=120` does not
> We hope to learn more about values in between those two, as we study the data more.

Reviewer: ```suggestion
This is a tradeoff between performance and freshness, and the cache allows site owners to provide SXGs with a max-age anywhere between 2 minutes and 7 days, to fit each page's particular needs. Anecdotally, we find that:

- `max-age=86400` (1 day) or longer works well for performance
- `max-age=120` does not

We hope to learn more about values in between those two, as we study the data more.
```

{% Aside %}

> We plan to make this tradeoff easier to manage in the future, by adding a cache purge API.  When that happens, it should be announced on [webpackaging-announce](https://groups.google.com/g/webpackaging-announce).

Reviewer: ```suggestion
We plan to make this tradeoff easier to manage in the future, by adding a cache purge API. When that happens, it should be announced on [webpackaging-announce](https://groups.google.com/g/webpackaging-announce).
```

    {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/gjfeKIq18vAAnZqEVkhj.png",
           alt="Network waterfall with SXG prefetch; LCP is 1.3 seconds",
           width="680", height="1026" %}
  </web-tab>
</web-tabs>

To diagnose this, we looked at the film strips. We found that the page rendered differently in SXG. In plain HTML, Chrome determined that the "largest element" for LCP was the headline. However, in the SXG version, the page added a lazy-loaded banner, which pushed the headline below the fold and caused the new largest element to be the lazy-loaded cookie consent dialog. Everything rendered faster than before, but a change in layout caused the metric to report it as slower.

We dug deeper and discovered the reason for this difference is that the page varies by `User-Agent`, and there was an error in the logic. It was serving a desktop page even though the SXG crawl header indicated mobile. After this was fixed, the browser correctly identified the page's headline as its largest element again.

Right now, SXGs are only enabled for Android, but we will soon enable them for desktop OSes. To prepare for this future, ensure that one of these is true:

>   - Your page doesn't `Vary: User-Agent` (e.g. it uses [responsive design](https://developers.google.com/search/mobile-sites/mobile-seo/responsive-design) or [separate mobile/desktop URLs](https://developers.google.com/search/mobile-sites/mobile-seo/separate-urls)).
>   - If it uses [dynamic serving](https://developers.google.com/search/mobile-sites/mobile-seo/dynamic-serving), it annotates itself as mobile- or desktop-only using [`<meta name=supported-media content=...>`](https://github.com/google/webpackager/blob/main/docs/supported_media.md).

Reviewer: ```suggestion
- Your page doesn't `Vary: User-Agent` (e.g. it uses [responsive design](https://developers.google.com/search/mobile-sites/mobile-seo/responsive-design) or [separate mobile/desktop URLs](https://developers.google.com/search/mobile-sites/mobile-seo/separate-urls)).
- If it uses [dynamic serving](https://developers.google.com/search/mobile-sites/mobile-seo/dynamic-serving), it annotates itself as mobile- or desktop-only using [`<meta name=supported-media content=...>`](https://github.com/google/webpackager/blob/main/docs/supported_media.md).
```

> ### subresources

Reviewer: ```suggestion
### Subresources
```

> SXG can be used to prefetch subresources (including images) along with the HTML. Cloudflare ASX will scan the HTML for same-origin (first-party) `<link rel=preload>` tags and convert them into [SXG-compatible Link headers](https://github.com/WICG/webpackage/blob/main/explainers/signed-exchange-subresource-substitution.md). Details in the source code [here](https://github.com/google/sxg-rs/blob/d16e66a2b8d5e0c069ec70af198dd95f640321d8/cloudflare_worker/worker/src/index.ts#L250-L275) and [here](https://github.com/google/sxg-rs/blob/main/sxg_rs/src/link.rs).

Reviewer: ```suggestion
SXG can be used to prefetch subresources (including images) along with the HTML. Cloudflare ASX will scan the HTML for same-origin (first-party) `<link rel=preload>` elements and convert them into [SXG-compatible Link headers](https://github.com/WICG/webpackage/blob/main/explainers/signed-exchange-subresource-substitution.md). Details in the source code [here](https://github.com/google/sxg-rs/blob/d16e66a2b8d5e0c069ec70af198dd95f640321d8/cloudflare_worker/worker/src/index.ts#L250-L275) and [here](https://github.com/google/sxg-rs/blob/main/sxg_rs/src/link.rs).
```

If it's working, you'll see additional prefetches from Google Search:

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/e09y5ZGYE93zgPi2pyFk.png",
>        alt="Google Search results with DevTools Network tab, showing a prefetch of /sub/…/image.jpg",
>        width="800", height="493" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/e09y5ZGYE93zgPi2pyFk.png", alt="Google Search results with DevTools Network tab, showing a prefetch of /sub/…/image.jpg", width="800", height="493" %}
</figure>
```

### Server-side metrics

When measuring server-side metrics such as [Time to First Byte (TTFB)](https://web.dev/ttfb/), it's important to note that your site should only be serving SXGs to crawlers that accept the format. Limit your measurement of TTFB to requests coming from real users, and not bots. You may find that generating SXGs increases the TTFB for crawler requests, but this has no impact on your visitors' experience.

### Client-side metrics

SXGs produce the most speed benefit for client-side metrics; especially LCP. However, their biggest impact is limited to scenarios where they can be prefetched. When measuring their impact, you could simply enable Cloudflare ASX, wait for it to be re-crawled by Googlebot, wait an additional 28 days for Core Web Vitals (CWV) aggregation, and then look at your new CWV numbers. However, the change might be harder to spot when mixed in among all the other changes during this time frame.
 
Instead, I find it helpful to "zoom in" on the potentially affected page loads, and frame it as, "SXGs affect X% of page views, improving their LCP by Y milliseconds at the 75th percentile."
 

> Currently, SXG prefetch only happens under certain conditions:
>   - Chromium browser (e.g. Chrome or Edge except on [iOS](https://chromium.googlesource.com/chromium/src.git/+/HEAD/docs/ios/user_agent.md)), version M87 or higher
>   - Android OS
>   - `Referer: google.com` or other [Google search domains](https://source.chromium.org/chromium/chromium/src/+/main:components/google/core/common/google_tld_list.h;l=13-42;drc=bea0acf022da996a8ff1dbdee378667a66c768e4). (Note that in Google Analytics, a referral tag applies to [all page views in the session](https://support.google.com/analytics/answer/6205762?hl=en#flowchart&zippy=%2Cin-this-article), whereas SXG prefetch only applies to the first page view, directly linked from Google Search.)

Reviewer: ```suggestion
Currently, SXG prefetch only happens under certain conditions:

- Chromium browser (e.g. Chrome or Edge except on [iOS](https://chromium.googlesource.com/chromium/src.git/+/HEAD/docs/ios/user_agent.md)), version M87 or higher
- Android OS
- `Referer: google.com` or other [Google search domains](https://source.chromium.org/chromium/chromium/src/+/main:components/google/core/common/google_tld_list.h;l=13-42;drc=bea0acf022da996a8ff1dbdee378667a66c768e4). (Note that in Google Analytics, a referral tag applies to [all page views in the session](https://support.google.com/analytics/answer/6205762?hl=en#flowchart&zippy=%2Cin-this-article), whereas SXG prefetch only applies to the first page view, directly linked from Google Search.)
```

{% endAside %}
 
Read on for how to measure "X% of page views" and "improving their LCP by Y milliseconds", but note that this is an area I'm still learning more about. If you have additional advice on how to capture SXG performance in Real User Monitoring (RUM), please let us know! [File a bug against developer.chrome.com](https://github.com/GoogleChrome/developer.chrome.com) with your suggested improvements.

### Contemporary study

When looking at production data, you can split page loads into SXG and non-SXG. Cloudflare ASX provides support for [HTML syntaxes](https://github.com/google/sxg-rs#sxg-only-behavior) that let you determine in JS whether the page load is SXG.

However, it is essential to limit the set of page loads you look at, so that the non-SXG side matches the eligibility conditions for SXG, in order to avoid selection bias. Otherwise, all of the following would exist *only* in the set of non-SXG page loads, which may have innately different LCP:

>   - iOS devices — due to differences in hardware or network speed among the users who have these devices.
>   - older Chromium browsers — for the same reasons.
>   - desktop devices — for the same reasons or because the page layout causes a different "largest element" to be chosen.
>   - same-site navigations (visitors following links within the site) — because they can reuse cached subresources from the previous page load.

Reviewer: ```suggestion
- iOS devices — due to differences in hardware or network speed among the users who have these devices.
- older Chromium browsers — for the same reasons.
- desktop devices — for the same reasons or because the page layout causes a different "largest element" to be chosen.
- same-site navigations (visitors following links within the site) — because they can reuse cached subresources from the previous page load.
```

In Google Analytics (UA), [create two custom dimensions](https://support.google.com/analytics/answer/2709829) with scope "Hit", one named "isSXG" and one named "referrer". (The built-in "Source" dimension has [session scope](https://support.google.com/analytics/answer/2709828?hl=en#example-session&zippy=%2Cin-this-article), so it doesn't exclude same-site navigations.)

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/GartToDEdIzMEXN8sktm.png",
>        alt="Google Analytics dimension editor with recommended settings",
>        width="619", height="432" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/GartToDEdIzMEXN8sktm.png", alt="Google Analytics dimension editor with recommended settings", width="619", height="432" %}
</figure>
```

Create a custom segment named "SXG counterfactual" with the following filters ANDed together:

>   - **referrer** starts with **https://www.google.**
>   - **Browser** exactly matches **Chrome**
>   - **Browser** Version matches regex **^(8[7-9]|9[0-9]|[0-9]{3})**
>   - **Operating System** exactly matches **Android**
>   - **isSXG** exactly matches **false**

Reviewer: ```suggestion
- **referrer** starts with **https://www.google.**
- **Browser** exactly matches **Chrome**
- **Browser** Version matches regex **^(8[7-9]|9[0-9]|[0-9]{3})**
- **Operating System** exactly matches **Android**
- **isSXG** exactly matches **false**
```

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/tqpW8Jy95f1jYJeV5Opx.png",
>        alt="Google Analytics segment editor with recommended filters",
>        width="800", height="441" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/tqpW8Jy95f1jYJeV5Opx.png", alt="Google Analytics segment editor with recommended filters", width="800", height="441" %}
</figure>
```

> Create a copy of this segment, named "SXG", except with **isSXG** exactly matches **true**.

Reviewer: ```suggestion
Create a copy of this segment, named "SXG", except with `isSXG` exactly matches `true`.
```

In your site template, add the following snippet above the Google Analytics snippet. When generating an SXG, ASX will change `false` to `true`:

```html
<script data-issxg-var>window.isSXG=false</script>
```

> Customize your Google Analytics reporting script [as recommended](https://github.com/GoogleChrome/web-vitals#send-the-results-to-google-analytics) to record LCP. If you're using gtag.js, modify the  the 'config' command to set the custom dimension (replacing `dimension1` and `dimension2` with the names that Google Analytics says to use):

Reviewer: ```suggestion
Customize your Google Analytics reporting script [as recommended](https://github.com/GoogleChrome/web-vitals#send-the-results-to-google-analytics) to record LCP. If you're using gtag.js, modify the `'config'` command to set the custom dimension (replacing `'dimension1'` and `'dimension2'` with the names that Google Analytics says to use):
```

```js
gtag('config', 'YOUR_TRACKING_ID', {
  'dimension1': String(isSXG),
  'dimension2': document.referrer,
});
```

> If you're using analytics.js, modify the 'create' command as [documented here](https://developers.google.com/analytics/devguides/collection/analyticsjs/creating-trackers).

Reviewer: ```suggestion
If you're using analytics.js, modify the `'create'` command as [documented here](https://developers.google.com/analytics/devguides/collection/analyticsjs/creating-trackers).
```

After waiting a few days to collect some data, go to the Google Analytics Events report and add a drilldown for the SXG segment. This should fill in the X for "SXGs affect X% of page views":

> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/MNaYo8ZjOxto57txJvwf.png",
>        alt="Google Analytics Events report with SXG segment, showing 12.5% Unique Events",
>        width="363", height="111" %}
> 
> Finally, go to the [Web Vitals Report](https://web-vitals-report.web.app/), select "Choose segments", and select "SXG counterfactual" and "SXG".
> 
> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/tIsSMUofs9lMv3C0ncwe.png",
>        alt="Web Vitals Report with selections for SXG counterfactual and SXG",
>        width="657", height="164" %}
> 
> Click "Submit", and you should see LCP distributions for the two segments. This should fill in the Y for "improving their LCP by Y milliseconds at the 75th percentile":
> 
> {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/LTiuLf05jVOH2xe9RMOg.png",
>        alt="Web Vitals Report showing LCP distributions for SXG counterfactual and SXG",
>        width="800", height="363" %}

Reviewer: ```suggestion
<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/MNaYo8ZjOxto57txJvwf.png", alt="Google Analytics Events report with SXG segment, showing 12.5% Unique Events", width="363", height="111" %}
</figure>

Finally, go to the [Web Vitals Report](https://web-vitals-report.web.app/), select "Choose segments", and select "SXG counterfactual" and "SXG".

<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/tIsSMUofs9lMv3C0ncwe.png", alt="Web Vitals Report with selections for SXG counterfactual and SXG", width="657", height="164" %}
</figure>

Click "Submit", and you should see LCP distributions for the two segments. This should fill in the Y for "improving their LCP by Y milliseconds at the 75th percentile":

<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/LTiuLf05jVOH2xe9RMOg.png", alt="Web Vitals Report showing LCP distributions for SXG counterfactual and SXG", width="800", height="363" %}
</figure>
```

#### Caveats

Once you've applied all of the above filters, SXG counterfactual page loads should consist of things like these:

>   - cache misses — If the Google SXG Cache doesn't have a fresh copy of the SXG for a given URL, it will redirect to the original URL at your site.
>   - other result types — As of writing, Google Search only supports SXG for standard web results (aka "10 blue links") and a few other types. Others, like Top Stories Carousel, will link to the original URL at your site.
>   - ineligible URLs — If some pages on your site are not eligible for SXG (e.g. because they are not cacheable), they could appear in this set.

Reviewer: ```suggestion
- cache misses — If the Google SXG Cache doesn't have a fresh copy of the SXG for a given URL, it will redirect to the original URL at your site.
- other result types — As of writing, Google Search only supports SXG for standard web results (aka "10 blue links") and a few other types. Others, like Top Stories Carousel, will link to the original URL at your site.
- ineligible URLs — If some pages on your site are not eligible for SXG (e.g. because they are not cacheable), they could appear in this set.
```

There may be remaining bias between the SXG page loads and the above set of non-SXG page loads, but I believe it will be smaller in magnitude than the above biases. For instance, perhaps your non-cacheable pages are slower or faster than your cacheable pages. If you suspect this could be an issue, consider looking at the data limited to a specific SXG-eligible URL to see if its results match the overall study.

If your site has some AMP pages, they probably won't see performance improvements from enabling SXG, as they can already be prefetched from Google Search. Consider adding a filter to exclude such pages, to further "zoom in" on the relevant changes.

Lastly, even addressing all selection biases, there is risk that survivorship bias makes LCP improvements look like degradations in RUM statistics. [This article](https://simonhearne.com/2022/survorship-bias-in-webperf/) does a great job of explaining that risk, and suggests looking at some form of abandonment metric to detect whether this is happening.

### Before/after study

> To corroborate results from the contemporary study, it may be useful to do a comparison of LCP before and after enabling SXG. Don't limit to SXG page views, to eliminate the potential biases noted above. Instead, look at SXG-eligible results — the above segment definitions but without the **isSXG** constraint.

Reviewer: ```suggestion
To corroborate results from the contemporary study, it may be useful to do a comparison of LCP before and after enabling SXG. Don't limit to SXG page views, to eliminate the potential biases noted above. Instead, look at SXG-eligible results—the above segment definitions but without the `isSXG` constraint.
```

Note that Google Search may take up to several weeks to recrawl all pages on your site, in order to identify that SXG has been enabled for them. In those several weeks, there are other potential biases that may occur:

  - New browser releases or new device purchases may speed up page loads.
  - A significant event like a holiday may skew traffic from normal.

It also is helpful to look at overall 75th percentile LCP before and after, in order to corroborate the above studies. Learning about a subset of the population doesn't necessarily tell us about the overall population. For instance, let's say SXG improves 10% of page loads by 800ms.

>   - If these were already the 10% fastest page loads, then it won't affect 75p at all.
>   - If they were the 10% slowest page loads, but they were more than 800ms slower than the 75p LCP to begin with, then it won't affect 75p at all.

Reviewer: ```suggestion
- If these were already the 10% fastest page loads, then it won't affect 75p at all.
- If they were the 10% slowest page loads, but they were more than 800ms slower than the 75p LCP to begin with, then it won't affect 75p at all.
```

These are extreme examples, likely not reflective of reality, but hopefully illustrate the issue. In practice, I believe it's likely that SXG will affect the 75th percentile for most sites. Cross-site navigations tend to be some of the slowest, and improvements from prefetching tend to be significant.

### Opt-out some URLs

Lastly, one way to compare SXG performance could be to disable SXG for some subset of URLs on your site. For instance, you could set a `CDN-Cache-Control: no-store` header to prevent Cloudflare ASX from generating an SXG.

My intuition is that this has a bigger risk of selection bias. For instance, it may make a big difference whether your site's home page or a similarly popular URL is selected into the control group or the experiment group.

### Holdback study

The ideal way to measure impact would be to conduct a holdback study, which has these properties:

>   - In the experiment group, some random fraction of page views that *would* be SXG are "held back", and served as non-SXG instead. This allows for an "apples-to-apples" comparison between equivalent users, devices, scenarios, and pages.
>   - Those held-back (aka counterfactual) page views are labeled as such in the analytics. This allows for a "zoomed-in" view of the data, where we can compare SXG page loads in the control to SXG counterfactuals in the experiment. This, in turn, reduces noise from the other page loads that would be unaffected by SXG prefetch.

Reviewer: ```suggestion
- In the experiment group, some random fraction of page views that *would* be SXG are "held back", and served as non-SXG instead. This allows for an "apples-to-apples" comparison between equivalent users, devices, scenarios, and pages.
- Those held-back (aka counterfactual) page views are labeled as such in the analytics. This allows for a "zoomed-in" view of the data, where we can compare SXG page loads in the control to SXG counterfactuals in the experiment. This, in turn, reduces noise from the other page loads that would be unaffected by SXG prefetch.
```

</figure>

In many cases, checking the box to enable this feature is enough to get the kind of substantial improvement shown above. Sometimes, there are a few more steps to ensure these SXGs are working as intended at each stage of the pipeline, and to optimize pages to take full advantage of prefetch.

In the past couple of months since Cloudflare's launch, I've been reading and responding to questions on [various](https://support.google.com/webmasters/threads?hl=en&thread_filter=(%27%22signed%20exchange%22%27)) [forums](https://community.cloudflare.com/t/automatic-signed-exchanges-sxgs-beta-launch/312634) and learning how to advise sites on how to make sure they're getting the most out of their SXG deployments. This post is a collection of my advice. I'll walk through the steps to analyze SXG performance, find ways to improve it, and finally measure its real world impact.

## Introduction

An SXG is a file containing a URL, a set of HTTP response headers, and a response body—all cryptographically signed by a Web PKI certificate. When the browser loads an SXG, it verifies all of these:

> - the SXG hasn't expired
> - the signature matches the URL, headers, body, and certificate
> - the certificate is valid and matches the URL

Reviewer: ```suggestion
- The SXG hasn't expired
- The signature matches the URL, headers, body, and certificate
- The certificate is valid and matches the URL
```

<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/zUqSkRSLNkF26gjyymEo.png", alt="SXG Validator showing a check mark (✅) and a Content Type of application/signed-exchange;v=b3", width="600", height="281" %}
</figure>

{% Aside %}
SXGs are meant for SXG-capable crawlers, to enable prefetching. When users visit your site directly, you should not serve SXGs to them, since they will not be prefetched—doing so only adds overhead. Cloudflare ASX performs this [content negotiation](https://github.com/google/webpackager/blob/main/cmd/webpkgserver/README.md#content-negotiation) automatically.
{% endAside %}

The extension fetches the current URL with an `Accept` request header that says it prefers the SXG version. If you see a check mark (✅) next to Origin, that means an SXG was returned; you can skip to the [Indexing](#indexing) section.

If you see a cross mark (❌), that means an SXG wasn't returned:

<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/4T3la9HtoRDz8SK7lWGi.png", alt="SXG Validator showing a cross mark (❌) and a Content Type of text/html", width="483", height="162" %}
</figure>

> If Cloudflare ASX is enabled, then the most likely reason for a cross mark (❌) is because a cache control response header prevents it. ASX looks at headers with the following names:

Reviewer: ```suggestion
If Cloudflare ASX is enabled, then the most likely reason for a cross mark (❌) is because a cache control response header prevents it. ASX looks at headers with the following names:

```

{% Aside %}
These restrictions can be seen in more detail in the ASX code. For instance, cache control headers are inspected by the code that [determines when to sign](https://github.com/google/sxg-rs/blob/d16e66a2b8d5e0c069ec70af198dd95f640321d8/sxg_rs/src/headers.rs#L105-L116) and [what the signature duration should be](https://github.com/google/sxg-rs/blob/d16e66a2b8d5e0c069ec70af198dd95f640321d8/sxg_rs/src/headers.rs#L252-L272). Other headers are removed before signing [by this configuration](https://github.com/google/sxg-rs/blob/130fec932e100b1f9949903bc6c5a61452b9f042/cloudflare_worker/wrangler.example.toml#L42) or [by this code](https://github.com/google/sxg-rs/blob/d16e66a2b8d5e0c069ec70af198dd95f640321d8/sxg_rs/src/headers.rs#L279-L292). Other headers [prevent signing](https://github.com/google/sxg-rs/blob/d16e66a2b8d5e0c069ec70af198dd95f640321d8/sxg_rs/src/headers.rs#L325-L345) if not otherwise stripped.
{% endAside %}

> Another possible reason is the presence of a `Vary: Cookie` response header. Googlebot fetches SXGs without user credentials and may serve them to multiple visitors. If you serve different HTML to different users based on their cookie, then they could see an incorrect experience such as a logged out view.

Reviewer: ```suggestion
Another possible reason is the presence of a [`Vary: Cookie`](https://developer.mozilla.org/docs/Web/HTTP/Headers/Vary) response header. Googlebot fetches SXGs without user credentials and may serve them to multiple visitors. If you serve different HTML to different users based on their cookie, then they could see an incorrect experience such as a logged out view.
```

{% Aside %}
In the future, pages using `Vary: Cookie` will be supported by SXG, by being shown only to visitors without cookies for your site. This is [implemented in Chromium](https://crbug.com/1250532), and will be enabled in Google Search once that has rolled out more fully.
{% endAside %}

> Alternatively to the Chrome extension, you can use a tool like `curl`:

Reviewer: ```suggestion
Alternatively to the Chrome extension, you can use a tool like [`curl`](https://curl.se/):
```

<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/MNaYo8ZjOxto57txJvwf.png", alt="Google Analytics Events report with SXG segment, showing 12.5% Unique Events", width="363", height="111" %}
</figure>

Finally, go to the [Web Vitals Report](https://web-vitals-report.web.app/), select "Choose segments", and select "SXG counterfactual" and "SXG".

<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/tIsSMUofs9lMv3C0ncwe.png", alt="Web Vitals Report with selections for SXG counterfactual and SXG", width="657", height="164" %}
</figure>

Click "Submit", and you should see LCP distributions for the two segments. This should fill in the Y for "improving their LCP by Y milliseconds at the 75th percentile":

<figure>
  {% Img src="image/rULxC7pPw3PFS4o9xr7v8isFmCv1/LTiuLf05jVOH2xe9RMOg.png", alt="Web Vitals Report showing LCP distributions for SXG counterfactual and SXG", width="800", height="363" %}
</figure>

#### Caveats

Once you've applied all of the above filters, SXG counterfactual page loads should consist of things like these:

> - cache misses—If the Google SXG Cache doesn't have a fresh copy of the SXG for a given URL, it will redirect to the original URL at your site.
> - other result types—Currently, Google Search only supports SXG for standard web results and a few other types. Others, like Featured Snippets and Top Stories Carousel, will link to the original URL at your site.
> - ineligible URLs—If some pages on your site are not eligible for SXG (e.g. because they are not cacheable), they could appear in this set.

Reviewer: ```suggestion
- **Cache misses:** If the Google SXG Cache doesn't have a fresh copy of the SXG for a given URL, it will redirect to the original URL at your site.
- **Other result types:** Currently, Google Search only supports SXG for standard web results and a few other types. Others, like Featured Snippets and Top Stories Carousel, will link to the original URL at your site.
- **Ineligible URLs:** If some pages on your site are not eligible for SXG (e.g. because they are not cacheable), they could appear in this set.
```

### Before/after study

To corroborate results from the contemporary study, it may be useful to do a comparison of LCP before and after enabling SXG. Don't limit to SXG page views, to eliminate the potential biases noted above. Instead, look at SXG-eligible results—the above segment definitions but without the `isSXG` constraint.

Note that Google Search may take up to several weeks to recrawl all pages on your site, in order to identify that SXG has been enabled for them. In those several weeks, there are other potential biases that may occur:

- New browser releases or improvements in users' hardware may speed up page loads.
- A significant event like a holiday may skew traffic from normal.

It also is helpful to look at overall 75th percentile LCP before and after, to confirm the above studies. Learning about a subset of the population doesn't necessarily tell us about the overall population. For instance, let's say SXG improves 10% of page loads by 800ms.

> - If these were already the 10% fastest page loads, then it won't affect 75p at all.
> - If they were the 10% slowest page loads, but they were more than 800ms slower than the 75p LCP to begin with, then it won't affect 75p at all.

Reviewer: ```suggestion
- If these were already the 10% fastest page loads, then it won't affect the 75th percentile at all.
- If they were the 10% slowest page loads, but they were more than 800ms slower than the 75th percentile LCP to begin with, then it won't affect the 75th percentile at all.
```
